tests/e2e/ctl_v3_member_test.go

```diff
@@ -304,3 +312,229 @@ func TestRemoveNonExistingMember(t *testing.T) {
+
+			epc, promotedMembers := mustCreateNewClusterByPromotingMembers(t, e2e.LastVersion, 2, e2e.WithSnapshotCount(tc.snapshotCount))
+			defer func() {
+				require.NoError(t, epc.Close())
+			}()
+			require.Len(t, promotedMembers, 1)
+			t.Logf("Promoted member: %+v", promotedMembers[0])
+
+			if tc.snapshotCount != 0 {
+				t.Logf("Write %d keys to trigger a snapshot", tc.snapshotCount)
+				for i := 0; i < int(tc.snapshotCount); i++ {
+					err := epc.Etcdctl().Put(ctx, fmt.Sprintf("key_%d", i), fmt.Sprintf("value_%d", i), config.PutOptions{})
+					require.NoError(t, err)
+				}
+			}
+
+			if tc.writeToV3StoreSuccess {
+				t.Log("Skip manually changing the already promoted member to a learner in v3store")
+			} else {
+				t.Logf("Stopping all members")
+				require.NoError(t, epc.Stop())
+
+				t.Log("Manually changing the already promoted member to a learner again in the v3store of all members to simulate the issue https://github.com/etcd-io/etcd/issues/19557")
+				promotedMembers[0].IsLearner = true
+				for i := 0; i < len(epc.Procs); i++ {
+					mustSaveMemberIntoBbolt(t, epc.Procs[i].Config().DataDirPath, promotedMembers[0])
+				}
+			}
+
+			t.Log("Upgrading the cluster")
+			currentVersion, err := e2e.GetVersionFromBinary(e2e.BinPath.Etcd)
+			require.NoError(t, err)
+			lastVersion, err := e2e.GetVersionFromBinary(e2e.BinPath.EtcdLastRelease)
+			require.NoError(t, err)
+			err = e2e.DowngradeUpgradeMembers(t, nil, epc, len(epc.Procs), false, lastVersion, currentVersion)
+			require.NoError(t, err)
+
+			for _, proc := range epc.Procs {
+				if tc.snapshotCount != 0 {
+					e2e.AssertProcessLogs(t, proc, "Syncing member in v3store")
+				} else {
+					if tc.writeToV3StoreSuccess {
+						e2e.AssertProcessLogs(t, proc, "ignore already promoted member")
+					} else {
+						e2e.AssertProcessLogs(t, proc, "Forcibly apply member promotion request")
+					}
+				}
+			}
+
+			t.Log("Checking all members are ready to serve client requests")
+			for i := 0; i < len(epc.Procs); i++ {
+				e2e.AssertProcessLogs(t, epc.Procs[i], e2e.EtcdServerReadyLines[0])
+			}
+
+			t.Log("Ensure all members in v3store are voting members again")
+			for i := 0; i < len(epc.Procs); i++ {
+				t.Logf("Stopping the member: %d", i)
+				require.NoError(t, epc.Procs[i].Stop())
+
+				t.Logf("Checking all members in member's backend store: %d", i)
+				ensureAllMembersFromV3StoreAreVotingMembers(t, epc.Procs[i].Config().DataDirPath)
+
+				t.Logf("Starting the member again: %d", i)
+				require.NoError(t, epc.Procs[i].Start(ctx))
+			}
+		})
+	}
+}
+
+func mustCreateNewClusterByPromotingMembers(t *testing.T, clusterVersion e2e.ClusterVersion, clusterSize int, opts ...e2e.EPClusterOption) (*e2e.EtcdProcessCluster, []*etcdserverpb.Member) {
+	require.GreaterOrEqualf(t, clusterSize, 1, "clusterSize must be at least 1")
+
+	ctx := context.Background()
+
+	t.Logf("Creating new etcd cluster - version: %s, clusterSize: %v", clusterVersion, clusterSize)
```

Comment:
the `clusterVersion` is empty when it's current version. we can remove version information from log

Reply:
yes, it's true. It's fine although not perfect. We know it's current version when it's empty string.

We can try to resolve it separately, i.e. https://github.com/etcd-io/etcd/pull/19640

Reply:
> we can remove version information from log

I tend to keep it as it's important info for debug.